# Limpieza del dataset de bienes raíces

Este es un conjunto de datos (dataset) reales que fue descargado usando técnicas de web scraping. El archivo contiene registros de **Fotocasa**, el cual es uno de los sitios más populares de bienes raíces en España. Contiene miles de datos de casas reales publicadas en la web www.fotocasa.com.

El dataset fue descargado hace algunos años y en ningún caso se obtuvo beneficio económico de ello.

Tu objetivo es extraer tanta información como sea posible con el conocimiento que tienes hasta ahora de ciencia de datos.

In [2]:
import pandas as pd

#### Ejercicio 00

Lee el dataset data/real_estate.csv e intenta visualizar la tabla (★☆☆)

In [3]:
# Este archivo CSV contiene puntos y comas en lugar de comas como separadores
df = pd.read_csv('../data/real_estate.csv', sep=';')
df

,Unnamed: 0,id_realEstates,isNew,realEstate_name,phone_realEstate,url_inmueble,rooms,bathrooms,surface,price,...,level4Id,level5Id,level6Id,level7Id,level8Id,accuracy,latitude,longitude,zipCode,customZone
0,1,153771986,False,ferrari 57 inmobiliaria,912177526.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,2.0,103.0,195000,...,0,0,0,0,0,0,"40,2948276786438","-3,44402412135624",NaN,NaN
1,2,153867863,False,tecnocasa fuenlabrada ferrocarril,916358736.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,1.0,NaN,89000,...,0,0,0,0,0,1,"40,28674","-3,79351",NaN,NaN
2,3,153430440,False,look find boadilla,916350408.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,2.0,2.0,99.0,390000,...,0,0,0,0,0,0,"40,4115646786438","-3,90662252135624",NaN,NaN
3,4,152776331,False,tecnocasa fuenlabrada ferrocarril,916358736.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,1.0,86.0,89000,...,0,0,0,0,0,0,"40,2853785786438","-3,79508142135624",NaN,NaN
4,5,153180188,False,ferrari 57 inmobiliaria,912177526.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,2.0,2.0,106.0,172000,...,0,0,0,0,0,0,"40,2998774864376","-3,45226301356237",NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15330,15331,153901377,False,infocasa consulting,911360461.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,2.0,1.0,96.0,259470,...,0,0,0,0,0,0,"40,45416","-3,70286",NaN,NaN
15331,15332,150394373,False,inmobiliaria pulpon,912788039.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,1.0,150.0,165000,...,0,0,0,0,0,0,"40,36652","-3,48951",NaN,NaN
15332,15333,153901397,False,tecnocasa torrelodones,912780348.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,4.0,2.0,175.0,495000,...,0,0,0,0,0,0,"40,57444","-3,92124",NaN,NaN
15333,15334,152607440,False,inmobiliaria pulpon,912788039.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,2.0,101.0,195000,...,0,0,0,0,0,0,"40,36967","-3,48105",NaN,NaN


In [25]:
df.columns

Index(['Unnamed: 0', 'id_realEstates', 'isNew', 'realEstate_name',
       'phone_realEstate', 'url_inmueble', 'rooms', 'bathrooms', 'surface',
       'price', 'date', 'description', 'address', 'country', 'level1',
       'level2', 'level3', 'level4', 'level5', 'level6', 'level7', 'level8',
       'upperLevel', 'countryId', 'level1Id', 'level2Id', 'level3Id',
       'level4Id', 'level5Id', 'level6Id', 'level7Id', 'level8Id', 'accuracy',
       'latitude', 'longitude', 'zipCode', 'customZone'],
      dtype='str')

## Trabajando con un DataFrame

#### Ejercicio 01

¿Cuál es la casa más cara del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada. Para visualizar el resultado utiliza un f-string. Por ejemplo:

```py
f'La casa más cara se encuentra en la dirección: {address} y su precio es {price} €'
```

In [26]:
max_price = df['price'].max()

house = df[df['price'] == max_price]

address = house['address'].iloc[0]
price = house['price'].iloc[0]

print(f'La casa más cara se encuentra en la dirección: {address} y su precio es {price} €')

La casa más cara se encuentra en la dirección: El Escorial y su precio es 8500000 €


#### Ejercicio 02

¿Cuál es la casa más barata del dataset? (★☆☆)

Imprime la dirección y el precio de la casa seleccionada utilizando f-string

In [27]:
min_price = df['price'].min()

house = df[df['price'] == min_price]

address = house['address'].iloc[0]
price = house['price'].iloc[0]

print(f'La casa más barata se encuentra en la dirección: {address} y su precio es {price} €')

La casa más barata se encuentra en la dirección: Parla y su precio es 0 €


#### Ejercicio 03

¿Cuál es la casa más grande del dataset? (★☆☆)

Imprime la dirección y el área de las casas seleccionadas utilizando f-string

In [28]:
max_surface = df['surface'].max()

house = df[df['surface'] == max_surface]

address = house['address'].iloc[0]
surface = house['surface'].iloc[0]

print(f'La casa más grande se encuentra en la dirección: {address} y su superficie es {surface} m2')

La casa más grande se encuentra en la dirección: Sevilla la Nueva y su superficie es 249000.0 m2


#### Ejercicio 04

¿Cuál es la casa más pequeña del dataset? (★☆☆)

In [29]:
min_surface = df['surface'].min()

house = df[df['surface'] == min_surface]

address = house['address'].iloc[0]
surface = house['surface'].iloc[0]

print(f'La casa más pequeña se encuentra en la dirección: {address} y su superficie es {surface} m2')

La casa más pequeña se encuentra en la dirección: Calle Amparo,  Madrid Capital y su superficie es 15.0 m2


#### Ejercicio 05.

¿El dataset contiene valores no admitidos (NAs)? (★☆☆)

- Muestra el nombre de las filas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.
- También muestra el nombre de las columnas seguidas por un booleano (`True` o `False`) según contengan o no contengan NAs.

In [31]:
# NAs por fila
print(df.isna().any(axis=1))

# NAs por columna
print(df.isna().any(axis=0))

0        True
1        True
2        True
3        True
4        True
         ... 
15330    True
15331    True
15332    True
15333    True
15334    True
Length: 15335, dtype: bool
Unnamed: 0          False
id_realEstates      False
isNew               False
realEstate_name      True
phone_realEstate     True
url_inmueble        False
rooms                True
bathrooms            True
surface              True
price               False
date                False
description          True
address             False
country             False
level1              False
level2              False
level3              False
level4               True
level5              False
level6               True
level7               True
level8               True
upperLevel          False
countryId           False
level1Id            False
level2Id            False
level3Id            False
level4Id            False
level5Id            False
level6Id            False
level7Id            False
level8Id     

#### Ejercicio 06.

Elimina los NAs del dataset, si aplica (★★☆)

Muestra las dimensiones del DataFrame original y del DataFrame después de las eliminaciones.

In [32]:
print(f'Dimensiones originales: {df.shape}')

df_clean = df.dropna(subset=['price', 'rooms', 'bathrooms', 'surface', 'address', 'level5', 'realEstate_name'])

print(f'Dimensiones después de eliminar NAs: {df_clean.shape}')

Dimensiones originales: (15335, 37)
Dimensiones después de eliminar NAs: (13626, 37)


#### Ejercicio 07

¿Cuantas poblaciones (columna level5) contiene el dataset? (★☆☆)

- Muestra una lista con los nombres de las poblaciones
- Muestra el total de las mismas

In [33]:
populations = df_clean['level5'].unique()

print(populations)

<StringArray>
[        'Arganda del Rey',      'Boadilla del Monte',
             'Fuenlabrada',     'Las Rozas de Madrid',
         ' Madrid Capital',    'Villaviciosa de Odón',
                   'Pinto',               'Valdemoro',
            'Navalcarnero',      'Pozuelo de Alarcón',
 ...
   'Villavieja del Lozoya',                'Canencia',
 'Cabanillas de la Sierra',     'Buitrago del Lozoya',
        'Fresno de Torote',              'Robregordo',
       'Pinilla del Valle',               'Rascafría',
              'La Hiruela',    'Montejo de la Sierra']
Length: 165, dtype: str


In [34]:
total_populations = df_clean['level5'].nunique()

print(f'El dataset contiene {total_populations} poblaciones.')

El dataset contiene 165 poblaciones.


#### Ejercicio 08

¿Cuál es la media de precios en la población (columna level5) de "Arroyomolinos (Madrid)"? (★★☆)

In [35]:
arroyomolinos = df_clean[df_clean['level5'] == 'Arroyomolinos (Madrid)']

mean_price = arroyomolinos['price'].mean()

print(f'El precio medio en Arroyomolinos (Madrid) es {mean_price} €')

El precio medio en Arroyomolinos (Madrid) es 300021.0775862069 €


#### Ejercicio 09.

¿Los precios promedios de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

- Muestra ambos promedios
- Escribe en una celda markdown una conclusión sobre ellos

In [36]:
valdemorillo = df_clean[df_clean['level5'] == 'Valdemorillo']
galapagar = df_clean[df_clean['level5'] == 'Galapagar']

mean_valdemorillo = valdemorillo['price'].mean()
mean_galapagar = galapagar['price'].mean()

print(f'Precio promedio de Valdemorillo: {mean_valdemorillo} €')
print(f'Precio promedio de Galapagar: {mean_galapagar} €')

Precio promedio de Valdemorillo: 374341.17647058825 €
Precio promedio de Galapagar: 379015.7397260274 €


Los precios promedio de Valdemorillo y Galapagar no son iguales. Galapagar presenta un precio promedio ligeramente superior al de Valdemorillo.

#### Ejercicio 10

¿Los promedios de precio por metro cuadrado (precio/m2) de "Valdemorillo" y "Galapagar" son iguales? (★★☆)

> Pista: Crea una nueva columna llamada `pps` (*price per square* o precio por metro cuadrado) y luego analiza los valores.

- Muestra ambos promedios de precio por metro cuadrado
- Escribe en una celda markdown una conclusión sobre ellos

In [38]:
df_clean['pps'] = df_clean['price'] / df_clean['surface']

In [39]:
valdemorillo = df_clean[df_clean['level5'] == 'Valdemorillo']
galapagar = df_clean[df_clean['level5'] == 'Galapagar']

mean_pps_valdemorillo = valdemorillo['pps'].mean()
mean_pps_galapagar = galapagar['pps'].mean()

print(f'Precio promedio por m2 de Valdemorillo: {mean_pps_valdemorillo} €')
print(f'Precio promedio por m2 de Galapagar: {mean_pps_galapagar} €')

Precio promedio por m2 de Valdemorillo: 1313.7408343709315 €
Precio promedio por m2 de Galapagar: 1573.875437320109 €


Los precios promedio por metro cuadrado de Valdemorillo y Galapagar no son iguales. Galapagar presenta un precio promedio por metro cuadrado superior al de Valdemorillo.

#### Ejercicio 11

¿Cuántas agencia de bienes raíces contiene el dataset? (★★☆)

- Muestra el valor obtenido.

In [40]:
total_agencies = df_clean['realEstate_name'].nunique()

print(f'El dataset contiene {total_agencies} agencias de bienes raíces.')

El dataset contiene 1698 agencias de bienes raíces.


#### Ejercicio 12

¿Cuál es la población (columna level5) que contiene la mayor cantidad de casas?(★★☆)

- Muestra la población y el número de casas.

In [41]:
houses_population = df_clean['level5'].value_counts()

population = houses_population.index[0]
number_houses = houses_population.iloc[0]

print(f'La población con mayor cantidad de casas es {population}, con {number_houses} casas.')

La población con mayor cantidad de casas es  Madrid Capital, con 6014 casas.


---

## Trabajando con un subconjunto del DataFrame

#### Ejercicio 13

Ahora vamos a trabajar con el "cinturón sur" de Madrid.

Haz un subconjunto del DataFrame original que contenga las siguientes poblaciones (columna level5): "Fuenlabrada", "Leganés", "Getafe", "Alcorcón" (★★☆)

> Pista: Filtra el DataFrame original usando la columna `level5` y la función `isin`.

In [42]:
south_belt = df_clean[df_clean['level5'].isin(
    ['Fuenlabrada', 'Leganés', 'Getafe', 'Alcorcón']
)]

south_belt

,Unnamed: 0,id_realEstates,isNew,realEstate_name,phone_realEstate,url_inmueble,rooms,bathrooms,surface,price,...,level5Id,level6Id,level7Id,level8Id,accuracy,latitude,longitude,zipCode,customZone,pps
3,4,152776331,False,tecnocasa fuenlabrada ferrocarril,916358736.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,1.0,86.0,89000,...,0,0,0,0,0,"40,2853785786438","-3,79508142135624",NaN,NaN,1034.883721
85,86,153152077,False,sinergical inmobiliaria,NaN,https://www.fotocasa.es/es/comprar/vivienda/le...,1.0,1.0,50.0,107000,...,0,0,0,0,1,"40,35059","-3,82693",NaN,NaN,2140.000000
94,95,153995577,False,viviendas365com,911226014.0,https://www.fotocasa.es/es/comprar/vivienda/le...,3.0,2.0,120.0,320000,...,0,0,0,0,0,"40,31933","-3,77574",NaN,NaN,2666.666667
109,110,153586414,False,area uno asesores inmobiliarios,912664081.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,3.0,142.0,425000,...,0,0,0,0,0,"40,3313411","-3,8313868",NaN,NaN,2992.957746
161,162,153440116,False,prolucisa,916350416.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,2.0,123.0,239500,...,0,0,0,0,0,"40,2744807718156","-3,79192471504211",NaN,NaN,1947.154472
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15275,15276,153903887,False,aliseda servicios de gestion inmobiliaria,911368198.0,https://www.fotocasa.es/es/comprar/vivienda/al...,3.0,1.0,78.0,138000,...,0,0,0,0,1,"40,31381","-3,83733",NaN,NaN,1769.230769
15291,15292,151697757,False,unipiso,912788631.0,https://www.fotocasa.es/es/comprar/vivienda/al...,3.0,2.0,110.0,279000,...,0,0,0,0,0,"40,3259051","-3,76318",NaN,NaN,2536.363636
15305,15306,153902389,False,jadein ferrero,914871639.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,2.0,85.0,170000,...,0,0,0,0,0,"40,2882193","-3,8098617",NaN,NaN,2000.000000
15322,15323,153871864,False,gestion comercial,911220662.0,https://www.fotocasa.es/es/comprar/vivienda/ma...,3.0,1.0,91.0,112000,...,0,0,0,0,0,"40,28282","-3,78892",NaN,NaN,1230.769231


#### Ejercicio 14

Calcula la media y la varianza de muestra para las siguientes variables: precio, habitaciones, superficie y baños (★★★)

> Debes usar el subset obtenido en la pregunta 13

- Crea y visualiza un diccionario con todos los valores

In [43]:
results = {
    'price': {
        'mean': south_belt['price'].mean(),
        'variance': south_belt['price'].var()
    },
    'rooms': {
        'mean': south_belt['rooms'].mean(),
        'variance': south_belt['rooms'].var()
    },
    'surface': {
        'mean': south_belt['surface'].mean(),
        'variance': south_belt['surface'].var()
    },
    'bathrooms': {
        'mean': south_belt['bathrooms'].mean(),
        'variance': south_belt['bathrooms'].var()
    }
}

results

{'price': {'mean': np.float64(222516.37692307692),
  'variance': np.float64(15257967738.540672)},
 'rooms': {'mean': np.float64(3.005128205128205),
  'variance': np.float64(0.7060070438761069)},
 'surface': {'mean': np.float64(111.91025641025641),
  'variance': np.float64(4292.972680293605)},
 'bathrooms': {'mean': np.float64(1.6358974358974359),
  'variance': np.float64(0.5732925183502847)}}

#### Ejercicio 15

¿Cuál es la casa más cara de cada población del cinturón sur de Madríd? (★★☆)

> Debes usar el subset obtenido en la pregunta 13

- Genera un DataFrame con esta información
- Muestra tanto la dirección como el precio de la casa seleccionada de cada población.
- Genera conclusiones en una celda markdown

In [44]:
fuenlabrada = south_belt[south_belt['level5'] == 'Fuenlabrada']
leganes = south_belt[south_belt['level5'] == 'Leganés']
getafe = south_belt[south_belt['level5'] == 'Getafe']
alcorcon = south_belt[south_belt['level5'] == 'Alcorcón']

max_fuenlabrada = fuenlabrada[fuenlabrada['price'] == fuenlabrada['price'].max()]
max_leganes = leganes[leganes['price'] == leganes['price'].max()]
max_getafe = getafe[getafe['price'] == getafe['price'].max()]
max_alcorcon = alcorcon[alcorcon['price'] == alcorcon['price'].max()]

In [45]:
most_expensive = pd.DataFrame({
    'Población': ['Fuenlabrada', 'Leganés', 'Getafe', 'Alcorcón'],
    'Dirección': [
        max_fuenlabrada['address'].iloc[0],
        max_leganes['address'].iloc[0],
        max_getafe['address'].iloc[0],
        max_alcorcon['address'].iloc[0]
    ],
    'Precio': [
        max_fuenlabrada['price'].iloc[0],
        max_leganes['price'].iloc[0],
        max_getafe['price'].iloc[0],
        max_alcorcon['price'].iloc[0]
    ]
})

most_expensive

,Población,Dirección,Precio
0,Fuenlabrada,"Calle de Paulo Freire, 5, Fuenlabrada",490000
1,Leganés,"Avenida Reina Sofía, Leganés",650000
2,Getafe,Getafe,1050000
3,Alcorcón,Alcorcón,950000


Entre las cuatro poblaciones analizadas, la vivienda más cara se encuentra en Getafe, con un precio de 1.050.000 €, seguida por Alcorcón con 950.000 €. Por otro lado, las viviendas más caras de Leganés y Fuenlabrada tienen precios de 650.000 € y 490.000 €, respectivamente.

#### Ejercicio 16

¿Qué puedes decir acerca del precio por metro cuadrado (precio/m2) entre los municipios de 'Getafe' y 'Alcorcón'? (★★☆)

> Debes usar el subset obtenido en la pregunta 13
>
> Pista: Crea una nueva columna llamada `pps` (price per square en inglés) y luego analiza los valores

In [46]:
getafe = south_belt[south_belt['level5'] == 'Getafe']
alcorcon = south_belt[south_belt['level5'] == 'Alcorcón']

mean_pps_getafe = getafe['pps'].mean()
mean_pps_alcorcon = alcorcon['pps'].mean()

print(f'Precio promedio por m2 de Getafe: {mean_pps_getafe} €')
print(f'Precio promedio por m2 de Alcorcón: {mean_pps_alcorcon} €')

Precio promedio por m2 de Getafe: 2066.314949251463 €
Precio promedio por m2 de Alcorcón: 2242.9410457200215 €


Al comparar el precio promedio por metro cuadrado de Getafe y Alcorcón, se observa que Alcorcón presenta un valor superior. El precio promedio por m² es de aproximadamente 2.242,94 € en Alcorcón y 2.066,31 € en Getafe.

## Conclusiones

#### Ejercicio 17

A partir del análisis del dataset se pudieron identificar diferencias importantes entre las viviendas y las distintas poblaciones. La vivienda más cara tiene un precio de 8.500.000 €, mientras que el precio mínimo registrado es de 0 €. En cuanto a la superficie, se encontraron valores desde 15 m² hasta 249.000 m².
El dataset contiene valores faltantes, por lo que se realizó una limpieza de las variables utilizadas en el análisis, pasando de 15.335 a 13.626 registros. Luego de esta limpieza se identificaron 165 poblaciones y 1.698 agencias inmobiliarias, siendo Madrid Capital la población con mayor cantidad de viviendas, con 6.014 registros.
También se observaron diferencias de precios entre poblaciones. Galapagar presenta tanto un precio promedio como un precio promedio por metro cuadrado superior a Valdemorillo. Por otra parte, al analizar el cinturón sur de Madrid, compuesto por Fuenlabrada, Leganés, Getafe y Alcorcón, Getafe presenta la vivienda de mayor precio entre las cuatro poblaciones, con 1.050.000 €. Sin embargo, al comparar el precio promedio por metro cuadrado entre Getafe y Alcorcón, Alcorcón presenta el valor más alto, con aproximadamente 2.242,94 €/m² frente a 2.066,31 €/m² en Getafe.